Overset intergrid boundary and hole search

jbakosi opened the pull request:
Overset intergrid boundary and hole search

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/quinoacomputing/quinoa/597)
<!-- Reviewable:end -->

jbakosi committed:
Add integrid overset bnd based on sideset instead of mesh blk id

jbakosi committed:
Add regression test exercising integrid bnd based on sidesets

jbakosi committed:
Merge branch 'develop' into m1

jbakosi committed:
Ensure all m2m tests can pass while incrementally adding new functionality to m2m

jbakosi committed:
Add hole search

This adds tagging OversetFE::m_uc with 2.0 values for holes on bg mesh
cut out by an overset mesh, specified by the intergrid surface. See
regression test for details, which also was updated since the m_uc
changes.

jbakosi committed:
Fix doc

jbakosi committed:
Fix comments of hole search aggregation code

jbakosi committed:
Add integral-based hole search for overset

The idea is that the surface integral \int v dot n dA will evaluate to
4 pi for a point inside the surface (independent of the surface) and to
zero outside where

v = [ (x-xp)/r^3, (y-yp)/r^3, (z-zp)/r^3 ]
r = sqrt[ (x-xp)^2 + (y-yp)^2 + (z-zp)^2 ]

where [xp,yp,zp] is the point in question and [x,y,z] is a point of the
surface element dA (the triangle centroid is used).

jbakosi committed:
Merge branch 'develop' into m4

adityakpandare reviewed (changes requested):
This is great work @jbakosi! Thank you! A couple of comments below.

> *[Reviewable](https://reviewable.io/reviews/quinoacomputing/quinoa/597#-:-NkRLHG84MlC_vpuu2Oc:b-ayw51e)* status: 0 of 24 files reviewed, 2 unresolved discussions (waiting on @jbakosi)

___
*[`src/Inciter/OversetFE.cpp` line 219 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/597#-NkRCsJR5LpL4aMap4Iy:-NkRCsJR5LpL4aMap4Iz:bulbk8t) ([raw file](https://github.com/quinoacomputing/quinoa/blob/32584406b20de6e908ddb42866b41f71caf0646f/src/Inciter/OversetFE.cpp#L219)):*
> ```cpp
>   }
> 
>   for (std::size_t e=0; e<inpoel.size()/4; e++) {
> ```

Can we invert this set of loops? i.e. loop over points in `bp` and flag elements surrounding these points using `esup`.
___
*[`src/Inciter/OversetFE.cpp` line 923 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/597#-NkRJSbl0WRYepiPc7NC:-NkRJSbl0WRYepiPc7ND:bkr2eg6) ([raw file](https://github.com/quinoacomputing/quinoa/blob/32584406b20de6e908ddb42866b41f71caf0646f/src/Inciter/OversetFE.cpp#L923)):*
> ```cpp
>           if (blid == 103) {
>             for (auto i : ndset) m_uc(i,iflag) = 1.0;
>           }
> ```

We should now be able to remove this hard coding, since `setupIntergridBoundaries()` sets up these flags automatically (not based on `blid==103` hardcoded.


<!-- Sent from Reviewable.io -->

jbakosi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/quinoacomputing/quinoa/597)* status: 0 of 24 files reviewed, 2 unresolved discussions (waiting on @adityakpandare)

___
*[`src/Inciter/OversetFE.cpp` line 219 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/597#-NkRCsJR5LpL4aMap4Iy:-NkZbgac0P275q-PQUkO:b-imfs9q) ([raw file](https://github.com/quinoacomputing/quinoa/blob/32584406b20de6e908ddb42866b41f71caf0646f/src/Inciter/OversetFE.cpp#L219)):*
<details><summary><i>Previously, adityakpandare (Aditya Pandare) wrote…</i></summary><blockquote>

Can we invert this set of loops? i.e. loop over points in `bp` and flag elements surrounding these points using `esup`.
</blockquote></details>

Now that I think about this more, I'm not sure if exchanging the loops is clearly better.

If the outer loop is over the set of already-tagged points (in bp), we now need to look in psup to see what edges are connected to the points in bp. That needs the data structure psup first, generated for the full mesh. This is not stored by default. We could store it and then that would be the price. There is always a price of course. Which one is better, only a measurement can decide.


<!-- Sent from Reviewable.io -->

jbakosi committed:
Hook up hole flag, experiment with moving overset test

jbakosi committed:
Update m2m test using set id for intergrid boundary and hole search

jbakosi commented:
@adityakpandare I think is ready to merge - after a final look. Thanks for the help with this.

adityakpandare reviewed (changes requested):
Thank you @jbakosi. The CI ran into the following problems:

1. The `asynclogic_oversetfe` tests fails with `Exception: Quiescence or another error detected`. I am sending the stack trace to your email presently.
2. All the builds (gnu, intel, release, debug) hang because one of the `asynclogic_oversetfe` tests deadlocks.
3. clang throws the following compile-time warnings (`-Werror`):
```
src/Inciter/Transporter.cpp:515:66: error: implicit conversion loses integer precision:
'unsigned long' to 'std::unordered_set<int, std::hash<int>, std::equal_to<int>, std::allocator<int>>::value_type' (aka 'int')
[-Werror,-Wshorten-64-to-32]

   for (const auto& ibs : ib) for (auto s : ibs) usedsets.insert( s );
                                                          ~~~~~~  ^
src/Inciter/Discretization.cpp:489:27: error: implicit conversion changes signedness:
'std::tuple_element<0, const std::pair<const int, std::vector<unsigned long, std::allocator<unsigned long>>>>::type' (aka 'const int')
to 'const std::unordered_set<unsigned long, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<unsigned long>>::key_type' (aka 'const unsigned long')
[-Werror,-Wsign-conversion]

             if ( is.count(setid) ) {
                     ~~~~~ ^~~~~

src/Inciter/OversetFE.cpp:187:19: error: implicit conversion changes signedness:
'std::tuple_element<0, const std::pair<const int, std::vector<unsigned long, std::allocator<unsigned long>>>>::type' (aka 'const int')
to 'const std::unordered_set<unsigned long, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<unsigned long>>::key_type' (aka 'const unsigned long')
[-Werror,-Wsign-conversion]

     if ( is.count(setid) ) {
             ~~~~~ ^~~~~

src/Inciter/OversetFE.cpp:288:15: error: unused variable 'ib' [-Werror,-Wunused-variable]

   const auto& ib = g_inputdeck.get< tag::param, tag::compflow,
               ^
```

Could you take a look at these?

> Reviewed 9 of 24 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/quinoacomputing/quinoa/597#-:-Nnj9hgc-AoOm8S11mgJ:by8anxw)* status: 9 of 42 files reviewed, 1 unresolved discussion (waiting on @jbakosi)



<!-- Sent from Reviewable.io -->

jbakosi committed:
Fix most oversetfe asynclogic tests

The following still deadlock:

  inciter:asynclogic_oversetfe_u0.0_migr_pe3
  inciter:asynclogic_oversetfe_u0.0_migr_pe8
  inciter:asynclogic_oversetfe_u0.5_migr_pe1
  inciter:asynclogic_oversetfe_u0.5_migr_pe3
  inciter:asynclogic_oversetfe_u0.5_migr_pe8
  inciter:asynclogic_oversetfe_u0.9_migr_pe1
  inciter:asynclogic_oversetfe_u0.9_migr_pe3
  inciter:asynclogic_oversetfe_u0.9_migr_pe8

i.e., those that actually migrate.

jbakosi commented:
The asynclogic problems (partially) should be taken care of by https://github.com/quinoacomputing/quinoa/pull/597/commits/7bdcd1e54fb31e6612a0f1260815e0dca9e33d21. Those that migrate still deadlock, you could disable migration for those by

```patch
diff --git a/tests/regression/inciter/transport/SlotCyl/asynclogic/CMakeLists.txt b/tests/regression/inciter/transport/SlotCyl/asynclogic/CMakeLists.txt
index 9ded28ef3..aba6eff55 100644
--- a/tests/regression/inciter/transport/SlotCyl/asynclogic/CMakeLists.txt
+++ b/tests/regression/inciter/transport/SlotCyl/asynclogic/CMakeLists.txt
@@ -27,6 +27,7 @@ function(add_async_test scheme virt npes ppn migr)
     set(ARG "-c slot_cyl_${scheme}.q -v -u ${virt} -b ${migr_cmd}")
     set(TEXT_RES "diag.0;diag.1")
     set(TEXT_BAS "noop.ndiff.cfg;noop.ndiff.cfg")
+    set(migr_cmd "")
   else()
     set(ARG "-c slot_cyl_${scheme}.q -i unitcube_1k.exo -v -u ${virt} -b ${migr_cmd}")
     set(TEXT_RES diag)
```

jbakosi committed:
Address some compiler warnings

jbakosi committed:
Add forgotten data to pup

Unfortunately, this still does not fix migrating overset asynclogic tests.

jbakosi commented:
I did find a couple of data structures in Discretization that I forgot to pup, fixed in https://github.com/quinoacomputing/quinoa/pull/597/commits/3d5a0300261d11d30f6d24dd47fc4524e8fb6632, but that still does not fix the migrating overset asynclogic tests. At this point I'm not sure how to fix those. Unless we come up with something obvious, I vote for disabling those tests, and putting migration to the overset todo list.

adityakpandare committed:
added isentropic vortex advection problem

jbakosi committed:
Skip setting up holes if overset is configured with  single mesh
